### PR TITLE
Checkly agent guide

### PR DESCRIPTION
**Pull request by brampling**

### CODE REVIEW

File: site/content/docs/private-locations/checkly-agent-guide.md [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+The Checkly Agent is a container-based tool that enables private locations in Checkly. The agent runs on your infrastructure in a container and runs checks on behalf of the Checkly application. Since the agent is local to your infrastructure, checks can run against any application accessible from the agent via HTTPS or HTTP.
+
+Here's how private locations work:
+
+A private location is similar to our existing public locations, but you have control over it. A private location can be in your own segregated network,  on-premises or cloud-based. You need one or more Checkly agents installed in that location/network with access to your applications. Checks are configured and scheduled in the Checkly cloud-based web UI as usual, but selecting a private location runs the check on the agent(s) in your infrastructure. Agents are stateless and ephemeral. As long as you have at least one operational agent, checks will run in your private location. Adding more agents will distribute the load and improve resilience and performance automatically.
+
+Here are the requirements before you get started:
```
Comment (umutuzgur) [inline site/content/docs/private-locations/checkly-agent-guide.md:23]: This part also exists in the "Getting started with private locations" page. This page feels more like "Optimizing Checkly Agent" rather than "Installing Checkly Agent" btw

**Comment by clample** [comment]
Is the plan to get rid of the older "Getting started with private locations" page?

File: site/content/docs/private-locations/checkly-agent-guide.md (hunk 1 above)
Comment (brampling) [reply site/content/docs/private-locations/checkly-agent-guide.md:23]: No, the plan is to keep both.

File: site/content/docs/private-locations/checkly-agent-guide.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+ ### Scale out
+
+To determine the number of agents you need in a private location you first need to know the number of checks assigned to the location and their frequency. API checks can be scheduled as frequently as every 10 seconds and browser checks as frequently as every 1 minute. Based on your configuration, you should be able to estimate how many checks will run per minute in your private location. Checks have a maximum running time of 30 seconds. This means that in the worst case scenario, a checkly agent with a `JOB_CONCURRENCY` of 1 can run two checks per minute. In an average configuration this will be higher and API checks are generally faster than browser checks.
+
+Once you have an idea of how many checks will be running and you know the per-agent `JOB_CONCURRENCY` limit based on the agent container memory allocation, you can estimate the number of agents required as `agents = (JOB_CONCURRENCY / Checks per minute / 2) + 1`. This should give you a safe amount of overhead to run your checks and the `+ 1` is to ensure redundant capactity in case of rolling upgrades or an agent failure. 
+
+## Installing and configuring the Checkly agent:
```
Comment (clample) [inline site/content/docs/private-locations/checkly-agent-guide.md:61]: Maybe it makes sense to put this before the "Redundancy and scaling" section?

I think users will need to get the Checkly Agent running before they worry about scaling and redundancy.

**Comment by umutuzgur** [comment]
@brampling I just realized that we don't mention how the pricing works. We should mention that the check runs are added just like the public locations

**Comment by brampling** [comment]
> @brampling I just realized that we don't mention how the pricing works. We should mention that the check runs are added just like the public locations

We don't mention pricing anywhere in the docs. The pricing page quotes price per browser check and per API check, that doesn't change.

File: site/content/docs/private-locations/checkly-agent-guide.md (hunk 1 above)
Comment (clample) [reply site/content/docs/private-locations/checkly-agent-guide.md:23]: I think "Installing Checkly Agent" and "Getting Started with Private Locations" are pretty similar topics, and the two pages seem to have quite a bit of overlap.

I think that it would make sense to restructure things and have a more clear distinction between all of the pages.

File: site/content/docs/private-locations/checkly-agent-guide.md [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+title: Installing Checkly Agent
+weight: 52
+menu:
+  docs:
+    parent: "Private Locations"
+aliases:
+- "/docs/private-locations/checkly-agent-guide/"
+---
+
+Note: The Checkly Agent currently only supports x86/AMD64 architecture and not ARM64. This means Apple Mac M1 systems are not currently supported.
```
Comment (clample) [inline site/content/docs/private-locations/checkly-agent-guide.md:11]:
```suggestion
{{<info>}}
Note: The Checkly Agent currently only supports x86/AMD64 architecture and not ARM64. This means Apple Mac M1 systems are not currently supported.
{{<info>}}
```

File: site/content/docs/private-locations/checkly-agent-guide.md [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+**You must have the Owner or Admin role on your Checkly account to create private locations.**
+
+The Checkly Agent is a container-based tool that enables private locations in Checkly. The agent runs on your infrastructure in a container and runs checks on behalf of the Checkly application. Since the agent is local to your infrastructure, checks can run against any application accessible from the agent via HTTPS or HTTP.
+
+Here are the requirements before you get started:
+
+- The API key received when creating a private location in the Checkly account settings page
+- Container runtime (we test using Docker, but other OCI-compliant runtimes should work)
+- Outbound internet access to https://agent.checklyhq.com (proxy configuration is supported)
+- Access to your internal API- or browser-based application
```
Comment (clample) [inline site/content/docs/private-locations/checkly-agent-guide.md:24]:
```suggestion
- Access to your internal API or browser-based application
```

File: site/content/docs/private-locations/checkly-agent-guide.md [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+## Redundancy and scaling
+
+We recommend at least two agents per private location for redundancy. When considering scaling, you should use at least N+1 agents per private location to ensure checks will be processed even if an agent fails. Agents are stateless, ephemeral, and scale automatically. You can add additional agents to a location at any time and remove them as necessary. If an agent fails, the other agents in the same private location will continue to run the checks assigned to that location. Even checks that are currently in progress on an agent that fails will be rerun by another agent after a timeout period (140s). There are two cases where checks assigned to a location could fail to run:
```
Comment (clample) [inline site/content/docs/private-locations/checkly-agent-guide.md:28]: > Agents are stateless, ephemeral, and scale automatically.

I'm not sure if it's accurate to say that the Agent scales automatically. We can't start more Agent containers on the customers behalf, so it's up to them to handle scaling.

Comment (brampling) [reply site/content/docs/private-locations/checkly-agent-guide.md:30]: Good point. Let me rephrase that a little.

File: site/content/docs/private-locations/checkly-agent-guide.md (hunk 3 above)
Comment (clample) [reply site/content/docs/private-locations/checkly-agent-guide.md:12]: It might make sense to do the same change on the "Getting started with private locations" page as well.

File: site/content/docs/private-locations/checkly-agent-guide.md [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+```
+[checkly-agent] Starting Consumer c7495186-6f1e-4526-b173-14ee9ad21775
+[checkly-agent] No jobs. Waiting..
+[checkly-agent] No jobs. Waiting..
+[checkly-agent] No jobs. Waiting..
+[checkly-agent] No jobs. Waiting..
+[checkly-agent] No jobs. Waiting..
+```
+
+8) Start additional agents as required. A single agent is fine for testing but multiple agent containers are recommended for redundancy. These agent containers should be on different hosts or nodes.
```
Comment (clample) [inline site/content/docs/private-locations/checkly-agent-guide.md:107]: > These agent containers should be on different hosts or nodes.

Maybe it's worth clarifying that this isn't a technical limitation of the Checkly Agent and that we only recommend this for redundancy / availability?

Something like:

> 8) Start additional agents as required. While single agent is fine for testing, we recommend running multiple agents across multiple hosts for high availability.

File: site/content/docs/private-locations/checkly-agent-guide.md [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+9) Refresh the private locations page in the Checkly app and you will see a count of the number of running agents.
+
+![agent running](/docs/images/private-locations/agent_running.png)
+
+## Updating the agent container
+
+Since the agents are stateless, they are updated by replacing them or updating the image in place. An in-place upgrade is easiest as it keeps the previously defined environment variables. You can use this command to do an in-place upgrade of an agent container. Ensure you have sufficient agent capacity as the agent container will have a short outage as it is upgraded. As agent shutdowns are graceful, no running checks will be lost:
+
+`docker run --rm -v /var/run/docker.sock:/var/run/docker.sock containrrr/watchtower --run-once <container-name-to-be-updated>`
```
Comment (clample) [inline site/content/docs/private-locations/checkly-agent-guide.md:117]: It might make sense to clarify that this is a third-party tool and link to https://containrrr.dev/watchtower/

I also think it could be more explicit that this is optional. Some users might prefer to just update the agent version in their IaC config file.

**Comment by clample** [comment]
Overall it looks good to me. 

I'm wondering if some of the more in-depth details (proxy support, automatic agent updates, key rotation, etc) will scare off some potential users who are less sophisticated and just want a simple setup. Maybe it would help to make a more clear split in describing the basic setup process (one agent, default job concurrency, etc), then describing the more sophisticated add-ons.

I think it's also fine as-is, though, and we can iterate as we go. I'll wait for @umutuzgur to check it out as well, though.

**Comment by umutuzgur** [comment]
@brampling @clample the markdown looks weird. I think there is a bug somewhere

**Comment by clample** [comment]
I'll check it out

**Review by umutuzgur — APPROVED** [review]
LGTM

File: site/content/docs/private-locations/checkly-agent-guide.md (hunk 7 above)
Comment (brampling) [reply site/content/docs/private-locations/checkly-agent-guide.md:117]: Fixed.

**Review by clample — DISMISSED** [review]

**Review by clample — APPROVED** [review]
